{"metadata":{"language_info":{"name":"bash","codemirror_mode":"shell","mimetype":"text/x-sh","file_extension":".sh"},"kernelspec":{"name":"bash","display_name":"Bash","language":"bash"}},"nbformat_minor":4,"nbformat":4,"cells":[{"cell_type":"markdown","source":"## Manipulating files with AWK\n### Bioinformatics Coffee Hour - April 14, 2020\n\n### What is awk?\nInvented in the 1970's, [awk](https://en.wikipedia.org/wiki/AWK) is a scripting language included in most Unix-like operating systems. It specializes in one-liner programs and manipulating text files.\n\nIn many cases, if you're parsing information from a text file (such as a [BED](https://en.wikipedia.org/wiki/BED_(file_format)) file, [FASTA](https://en.wikipedia.org/wiki/FASTA) file, etc.), you could write a Python script...or you could do it with awk in a single line!\n\n### Syntax\nawk scripts are organized as:\n\n`awk 'pattern { action; other action }' file`\n\nMeaning that every time that the pattern is true, awk will execute the action in the brackets.\nIf no pattern is specified, the action will be taken for every line in the input file, e.g. the following command prints every line:","metadata":{}},{"cell_type":"code","source":"awk '{print}' data/hg38.genome | head","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"The two most important patterns are `BEGIN` and `END`, which tell the action to take place before any lines are read and after the last line.","metadata":{}},{"cell_type":"code","source":"awk 'BEGIN{sum=0} {sum+=1} END {print sum}' data/hg38.genome","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":" The above line sets a variable at the start of the script, adds 1 to it every line, then prints its value at the end.","metadata":{}},{"cell_type":"markdown","source":"If a variable hasn't been initialized, it is treated as 0 in numeric expressions, and an empty string in string expressions&mdash;awk will not print an error!\nSo the following awk script also prints the number of lines in the file data/hg38.genome:","metadata":{}},{"cell_type":"code","source":"awk '{sum+=1} END {print sum}' data/hg38.genome","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"### Input and output\nInput to awk is split into **records** and **fields**.\n- By default, **records** are separated by newline character, i.e # of records = # of lines in input file\n- Each record is subdivided into **fields**, i.e. columns, as determined by the field separator (see below)\n\nThere are several important built-in variable in awk.\nThe fields (columns) of each record are referred to by `$number`, so the first column would be `$1`, second would be `$2`, etc. `$0` refers to the entire record.\n\nSo to print the second column of each line in the file, we'd use:","metadata":{}},{"cell_type":"code","source":"awk '{print $2}' data/hg38.genome | head","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"And if we wanted to print the second then the first:","metadata":{}},{"cell_type":"code","source":"awk '{print $2,$1}' data/hg38.genome | head","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Note that when the different fields are separated with commas in the `print` statement, they are joined by the output field separator (the **OFS** variable, described below), which is by default a space.\nIf the comma is omitted between fields (e.g., `awk '{print $2 $1}'`, they are concatenated without a separator.\n\nWe can also print strings using using quotation marks:","metadata":{}},{"cell_type":"code","source":"awk '{print \"First column:\" $1}' data/hg38.genome | head","metadata":{"trusted":true},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"Which for every line of the file will print the text \"First column:\" followed by the value in the first field.","metadata":{}},{"cell_type":"markdown","source":"---\nawk has several other built-in variables that are very useful for parsing text, including:\n\n|  |   |\n---|---|\n| **FS** | field separator (default: white space) |\n| **OFS** |  output field separator, i.e. what character separates fields when printing|\n| **RS** | record separator, i.e. what character records are split on (default: new line) |\n| **ORS** | output record separator |\n| **NR** | number of records in input (# lines by default) |\n\nAssigning to a field causes the entire record ($0) to be recomputed using **OFS**. We can use this to convert between file formats, e.g. make a comma-separated text file into a tab-separated file:","metadata":{}},{"cell_type":"code","source":"awk 'BEGIN{FS=\",\" ; OFS=\"\\t\"} {$1 = $1; print $0}' data/enhancers.csv > data/enhancers.tsv","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"### Conditionals and pattern matching\nLike other programming languages, awk allows conditional matching with if/else statements.\n\n`awk '{if(condition) action; else other_action}'`\n\nawk uses the following conditional operators:\n\n| | |\n|-|-|\n|==|equal to|\n|!=|not equal to|\n|>|greater than|\n|>=|greater than or equal to|\n|<|less than|\n|<=|less than or equal to|\n|&&|AND|\n| \\|\\| |OR|\n| ! | NOT |\n\nIn addition, awk also supports string matching using regular expressions, using the following expressions:\n\n| | |\n|-|-|\n|\\~|matches|\n|!~|does not match|\n\nFor string matching, the pattern being matched must be enclosed by slashes, like so:\n\n`awk '{if($1 ~ /pattern/) print}'`\n\nNote that if an action isn't specified, the default action is `{print}`, so the previous awk command is equivalent to the following, which specifies only a pattern expression:\n\n`awk '$1 ~ /pattern/'`\n\n---\n### Example uses:\n\n- Count number of sequences in a FASTQ file:","metadata":{}},{"cell_type":"code","source":"awk 'END{print NR/4}' data/Falb_COL2.1.fastq","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"**Note**: this is technically safer than the command we looked at earlier using grep, as yo don't have to worry about accidentally counting the quality line.\n\n- Only print annotations on a specific scaffold (chr1) that fall between 1Mb and 2Mb:","metadata":{}},{"cell_type":"code","source":"awk 'BEGIN{FS=\"\\t\";OFS=\"\\t\"} {if($1 == \"chr1\" && $2 >=1000000 && $2 <= 2000000) print}' data/ucscGenes.bed | head","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"**Note**: when we specify that we only want annotations from chr1, we're using exact match (`== \"chr1\"`) and not pattern match (`~ /chr1/`)...why is this??","metadata":{}},{"cell_type":"markdown","source":"- Only print lines of GFF file that match the string \"exon\" in their third column:","metadata":{}},{"cell_type":"code","source":"awk 'BEGIN{FS=\"\\t\"} {if($3 ~ /exon/) print $0}' data/dmel-all-no-analysis-r6.20.gff | head","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"- Convert from GFF (genome feature file) to BED file ","metadata":{}},{"cell_type":"code","source":"grep -v '^#' data/dmel-all-no-analysis-r6.20.gff | awk 'BEGIN{FS=\"\\t\"; OFS=\"\\t\"} {print $1,$4-1,$5}' | head","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"**Note**: remember that BED and GFF files have different coordinate systems, i.e. BED start coordinate is 0 based, half-open, GFF is 1-based inclusive! Also, we are first using grep to skip the header lines in the GFF file.\n\nAlternatively, you could do the whole thing with only awk, no grep required:","metadata":{}},{"cell_type":"code","source":"awk 'BEGIN{FS=\"\\t\"; OFS=\"\\t\"} !/^#/ {print $1,$4-1,$5}' data/dmel-all-no-analysis-r6.20.gff | head","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"With this command, `!/^#/` is a pattern (like `BEGIN` or `END`) that tell awk to execute the print statement when the start of the line does not match a `#`. Use whichever makes the most sense to you!\n\n### Practice\nUsing awk:\n\n* Pull out only the CDS annotations from the GFF file dmel-all-no-analysis-r6.20.gff and output them in BED format","metadata":{}},{"cell_type":"markdown","source":"*Try it*","metadata":{}},{"cell_type":"code","source":"awk '...' data/dmel-all-no-analysis-r6.20.gff | head","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"*Solution*","metadata":{}},{"cell_type":"code","source":"awk 'BEGIN{FS=\"\\t\"; OFS=\"\\t\"} {if($3 ~ /CDS/) print $1,$4-1,$5}' data/dmel-all-no-analysis-r6.20.gff","metadata":{"jupyter":{"source_hidden":true}},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"- Extract FASTA information from the SAM file Falb_COL2.final.sam (hint: you will need to remove the header lines first, they start with `@`!) (Another hint: sequence ID = 1st column, sequence = 10th column)","metadata":{}},{"cell_type":"markdown","source":"*Try it*","metadata":{}},{"cell_type":"code","source":"awk '...' data/Falb_COL2.final.sam | head","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"*Solution*","metadata":{}},{"cell_type":"code","source":"awk 'BEGIN{FS=\"\\t\"; OFS=\"\\n\"} !/^@/ {print \">\"$1,$10}' data/Falb_COL2.final.sam | head","metadata":{"jupyter":{"source_hidden":true}},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"- Write a command to calculate that average of the 5th column (i.e. mapping quality score) of a tab-separated SAM file Falb_COL2.final.sam and output it (hint: as above, need to remove headers)","metadata":{}},{"cell_type":"markdown","source":"*Try it*","metadata":{}},{"cell_type":"code","source":"awk '...' data/Falb_COL2.final.sam | head","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"*Solution*","metadata":{}},{"cell_type":"code","source":"awk 'BEGIN{FS=\"\\t\"; sum=0} !/^@/ {sum+=$5} END{print sum/NR}' data/Falb_COL2.final.sam | head","metadata":{"jupyter":{"source_hidden":true}},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"- Calculate the average length of gene annotations *only on the 2L arm* from the file dmel-genes.bed","metadata":{}},{"cell_type":"markdown","source":"*Try it*","metadata":{}},{"cell_type":"code","source":"awk '...' data/dmel-genes.bed","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":"*Solution*","metadata":{}},{"cell_type":"code","source":"awk 'BEGIN{FS=\"\\t\"; sum=0} $1 == \"2L\" {len=$3-$2; sum=sum+len} END{print sum/NR}' data/dmel-genes.bed","metadata":{"jupyter":{"source_hidden":true}},"execution_count":null,"outputs":[]}]}